### DISCUSSION TITLE
Simulating 3 Level Systems

### DISCUSSION THREAD

**Discussion by ArturDomingues** [qiskit-community/qiskit-dynamics#92] [Q&A] [answered]
Hello, I was trying to simulate some hyperfine interactions and I used qiskit nature to define S=1 operator, and when I tried to used the solver method it returned me and error saying that I couldn`t use matrices in that module, what can be done to simulate the interaction between a S=1/2 and S=1 particles? Do I need to define directly the hamiltonian in the S_total_z basis (S_total being the spin of the coupled system)?

**Comment by DanPuzzuoli** [comment, marked as answer]
Hi!

Do you mean that you tried using the matrices from qiskit nature in dynamics and it raised an error? If so what type are the matrices in qiskit nature? Can you share part of the error message?

**Reply by ArturDomingues** [reply]
I think I solved my problem. I was using x=SpinOp("X", spin=1) and not x=Operator(SpinOp("X", spin=1)), just this solved what was happening, but I really don't know why. so a explanation would be nice =D

**Reply by DanPuzzuoli** [reply]
Ah interesting. Taking a look at the class, the reason is that `SpinOp` (and more generally, the `SecondQuantizedOp` base class that it inherits from) doesn't have an `__array__` method. 

This is relevant because of the following:
- Internally, when you construct a `Solver` or one of the model classes in dynamics, it will attempt to convert all of the matrices you supply into arrays (either numpy, jax, or the `Array` class). This is true with whatever format you specify, whether they be specified as numpy arrays already, qiskit `Operator` instances, and even qutip `QObj` is supported.
- To convert an input matrix `x` to an array, we call `np.array(x)`. When calling `np.array(x)`: 
    - If `x` is something numpy recognizes (like a list of floats) it'll do its usual thing. 
    - If `x` however is *not* an instance of a class that numpy recognizes, it will check if `x` has an `__array__` method. If it does, calling `np.array(x)` will actually return `x.__array__()`. Basically numpy has provided a mechanism by which, if you write a new class, you can control how `np.array(x)` behaves.
    - If numpy does not recognize the type of `x`, AND `x` does not have an `__array__` method, an error is raised.
- That all being said, `Operator` *does* have an `__array__` method, whereas `SecondQuantizedOp` does not. Hence, if you pass operators in this format to dynamics it will fail.

(Note that I may not have the details of the operation of `np.array` 100% correct, but I think the general idea of what I'm saying is right.)

**Reply by ArturDomingues** [reply]
That explains the outputs of printing x as` SpinOp` and as `Operator`. I was having problems doing basic operations like `@` or raising x to a power because these aren't implemented in qiskit_nature, thanks for the nice explanation.
